### ISSUE TITLE
Normals / Modifier refactoring

### ISSUE THREAD

**Issue by julienduroure** [KhronosGroup/glTF-Blender-IO#877]
- [ ] How to detect if a modifier changes some normals ? See #860 and #874 
- [ ] See #799 
- [ ] #100 Custom normals in SK bug
- [ ] See #911

**Comment by julienduroure** [comment]
Added #100 to the list, because it seems to be linked to custom normals in SK

**Comment by julienduroure** [comment]
Added #911 to the list

**Comment by Joel-Hjerten** [comment]
I'm losing custom normals when using Apply Modifiers:
Blender 2.82, web viewer https://gltf-viewer.donmccurdy.com/
[Bug custom normals discarded by gltf export if apply modifiers.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/4288203/Bug.custom.normals.discarded.by.gltf.export.if.apply.modifiers.zip)
![explanation](https://user-images.githubusercontent.com/18649705/75896958-1a430200-5e38-11ea-8a1c-78ae011ba685.jpg)

**Comment by GilbertoRodrigues** [comment]
https://developer.blender.org/T45151

**Comment by scurest** [comment]
What happens if we delete the code that handles normals specially in __gather_mesh?

<details>

<summary>Like this</summary>

````diff
diff --git a/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_nodes.py b/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_nodes.py
index b09e7aa..208e2c7 100644
--- a/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_nodes.py
+++ b/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_nodes.py
@@ -254,12 +254,6 @@ def __gather_mesh(blender_object, library, export_settings):
     if blender_object.type != "MESH":
         return None
 
-    modifier_normal_types = [
-        "NORMAL_EDIT",
-        "WEIGHTED_NORMAL",
-        "BEVEL"
-    ]
-
     # If not using vertex group, they are irrelevant for caching --> ensure that they do not trigger a cache miss
     vertex_groups = blender_object.vertex_groups
     modifiers = blender_object.modifiers
@@ -269,16 +263,6 @@ def __gather_mesh(blender_object, library, export_settings):
         modifiers = None
 
     if export_settings[gltf2_blender_export_keys.APPLY]:
-        auto_smooth = blender_object.data.use_auto_smooth
-        edge_split = None
-        some_normals_modifier = any([m in modifier_normal_types for m in [mod.type for mod in blender_object.modifiers]])
-        if auto_smooth and not some_normals_modifier:
-            edge_split = blender_object.modifiers.new('Temporary_Auto_Smooth', 'EDGE_SPLIT')
-            edge_split.split_angle = blender_object.data.auto_smooth_angle
-            edge_split.use_edge_angle = not blender_object.data.has_custom_normals
-            blender_object.data.use_auto_smooth = some_normals_modifier
-            bpy.context.view_layer.update()
-
         armature_modifiers = {}
         if export_settings[gltf2_blender_export_keys.SKINS]:
             # temporarily disable Armature modifiers if exporting skins
@@ -298,10 +282,6 @@ def __gather_mesh(blender_object, library, export_settings):
             # restore Armature modifiers
             for idx, show_viewport in armature_modifiers.items():
                 blender_object.modifiers[idx].show_viewport = show_viewport
-
-        if auto_smooth and not some_normals_modifier:
-            blender_object.data.use_auto_smooth = True
-            blender_object.modifiers.remove(edge_split)
     else:
         blender_mesh = blender_object.data
         skip_filter = False

````

</details>

This fixes #975 and doesn't regress #860. I also looked in the OBJ exporter and noticed it didn't do anything like this.

**Comment by scurest** [comment]
I tested a model with auto smooth turned on (no custom normals). In blender-v2.83-release, with the above patch applied normals did export wrong. But in master the same model exported correctly even with that patch. I suspect its because #1127 changed how we get normals.

So I'm going to PR the patch.
